File: src/main/java/com/elchologamer/userlogin/command/base/SubCommand.java
package com.elchologamer.userlogin.command.base;

import org.bukkit.command.CommandSender;

public abstract class SubCommand extends BaseCommand {

    private final String permission;

    public SubCommand(String name) {
        this(name, null);
    }

    public SubCommand(String name, boolean playerOnly) {
        this(name, playerOnly, null);
    }

    public SubCommand(String name, String permission) {
        this(name, false, permission);
    }

    public SubCommand(String name, boolean playerOnly, String permission) {
        super(name, playerOnly);
        this.permission = permission;
    }

    @Override
    public final boolean run(CommandSender sender, String label, String[] args) {
        if (permission != null && !sender.hasPermission(permission)) {
            String message = this.getPermissionMessage();
            if (message != null) sender.sendMessage(message);
            return true;

        }

        return run(sender, args);
    }

    public abstract boolean run(CommandSender sender, String[] args);

    public String getPermission() {
        return permission;
    }
}

File: src/main/kotlin/com/elchologamer/userlogin/command/base/SubCommand.kt
package com.elchologamer.userlogin.command.base

import org.bukkit.command.CommandSender

abstract class SubCommand constructor(
    name: String,
    playerOnly: Boolean,
    private val permission: String? = null
) : BaseCommand(
    name, playerOnly
) {
    constructor(name: String, permission: String? = null) : this(name, false, permission)

    override fun run(sender: CommandSender, label: String, args: Array<String>): Boolean {
        if (permission?.let { !sender.hasPermission(it) } == true) {
            permissionMessage?.let { sender.sendMessage(it) }
            return true
        }
        return run(sender, args)
    }

    abstract fun run(sender: CommandSender, args: Array<String>): Boolean

    override fun getPermission(): String? {
        return permission
    }
}